### ISSUE TITLE
Framebuffer Emulation depends on GLES2 extension: OES_depth_texture

### ISSUE THREAD

**Issue by loganmc10** [gonetz/GLideN64#994]
I suppose this is a feature request more than a bug.

Currently Framebuffer Emulation uses depth texture attachments. For OpenGL ES2 that depends on this extension:

https://www.khronos.org/registry/gles/extensions/OES/OES_depth_texture.txt

For this reason, Framebuffer Emulation doesn't work on the Raspberry Pi, as it does not support the OES_depth_texture extension. See the following:

https://github.com/gonetz/GLideN64/issues/605

https://www.raspberrypi.org/forums/viewtopic.php?f=67&t=66247

It seems that using a renderbuffer instead of depth texture attachments would solve this issue, although I will admit I have no idea what work would be involved in that effort, I just thought I would bring this up as an issue.

**Comment by gonetz** [comment]
It is not hard. I used depth renderbuffer in earlier versions. I switched to texture because I need its content for depth buffer emulation. However, I need to think how to make it better without slicing code too much.

**Comment by gonetz** [comment]
@loganmc10 I've possibility to use renderbuffer instead of depth texture. Check commit 8f46834fb8843
Please test and report how it work.

**Comment by loganmc10** [comment]
Thanks! I'll test this out tomorrow. What would be a good game to test? And what sort of things should look for?

**Comment by loganmc10** [comment]
@gonetz I just tried this out with Framebuffer Emulation enabled, just a black screen (I can just hear sound).

I tried Super Mario 64

**Comment by loganmc10** [comment]
If I put "assert(!isGLError());" at the top of "void FrameBufferList::attachDepthBuffer()" I get:

```
glGetError 0x500
OpenGL Error: invalid enumerant
mupen64plus: /home/alarm/GLideN64/src/FrameBuffer.cpp:714: void FrameBufferList::attachDepthBuffer(): Assertion `!isGLError()' failed.
```

so I'll track down where exactly the glError is happening and let you know

**Comment by gonetz** [comment]
Incomplete Attachment: I had that error on GLES2 when size of depth renderbuffer differs from size of color buffer texture.

Super Mario 64 is a good choice to start testing that functionality. This game does not use fb efects, so frame and depth buffer sizes should be the same. May be Pi can't use texture as render target at all, even for color FBO attachment?

**Comment by loganmc10** [comment]
So I tracked down the source of the glError, and it is this line:

https://github.com/gonetz/GLideN64/commit/8f46834fb884395deba963d66a5002718c4ba006#diff-db8e829d1a9149099b17034ece42f036R184

```
glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT, m_depthRenderbufferWidth, height);
```

According to https://www.khronos.org/opengles/sdk/docs/man/xhtml/glRenderbufferStorage.xml

"GL_DEPTH_COMPONENT" isn't valid, but "GL_DEPTH_COMPONENT16" is. When I change it to be "GL_DEPTH_COMPONENT16", then it compiles and runs without any asserts failing or FBO errors, I'm not in front of my TV right now so I can't say what it looks like, but there are no more errors, I don't know what the difference between GL_DEPTH_COMPONENT and GL_DEPTH_COMPONENT16 is however.

**Comment by gonetz** [comment]
> "GL_DEPTH_COMPONENT" isn't valid, but "GL_DEPTH_COMPONENT16" is.

It is actually platform-dependent, check FBOTextureFormats::init()
If Raspberry Pi uses GL_DEPTH_COMPONENT16, no problem.

**Comment by loganmc10** [comment]
Ok I will test it out tonight, assuming Super Mario 64 works, what is a good game to test some Framebuffer Emulation features? Mario Tennis? Pokemon Stadium?

**Comment by gonetz** [comment]
@loganmc10 Since we correcting code for Raspberry Pi, I have a question/request: cmake defines VC (-DVC) for Raspberry Pi. It is very bad non-informative define for OS. I want to rename it to something more informative, OS_RASPBERRY_PI for example. Please suggest something better if you don't like it.

Mario Tennis is too complex for GLES2. It will be great success if you will get pause screen in Zelda OOT working correct. Lego Racers cars. Pause screen in Banjo Kazooie. Perfect Dark has plenty of fb effects, and some of them may work with GLES2.

**Comment by loganmc10** [comment]
Well VC is short for "VideoCore", the Raspberry Pi's GPU. I don't know of any other devices that use the VideoCore GPU, but presumably most of these issues are specific to the GPU, not the Raspberry Pi (since the GLES implementation is closed source and distributed by the chip manufacturer, Broadcom).

You could rename VC to OS_RASPBERRY_PI, doesn't really matter to me, but I assume if another device came out in the future that used the same GPU it would probably run into a lot (if not all) of these same issues.

**Comment by gonetz** [comment]
Thanks for the explanation. Ok, let it be VC, at least for now.

**Comment by loganmc10** [comment]
So this is what it looks like (Smash Bros, Zelda OOT, and Super Mario respectively, these are save states that I had)

![snapshot](https://cloud.githubusercontent.com/assets/848146/15656483/fddc73f8-2664-11e6-8da5-4eabfef5f381.png)
![snapshot2](https://cloud.githubusercontent.com/assets/848146/15656484/fe0715ea-2664-11e6-9f9f-f08231583081.png)
![snapshot3](https://cloud.githubusercontent.com/assets/848146/15656485/fe1cb9ae-2664-11e6-9db7-2fb1dc150c00.png)

**Comment by loganmc10** [comment]
You can't tell from the Mario screenshot but it flickers between the proper image and graphical corruption similar to this image:

https://cloud.githubusercontent.com/assets/10704701/14781877/b8020994-0b28-11e6-9620-f001fd2ff1c5.jpg

Eventually the whole display becomes corrupt. Golden Eye has the same issue as Zelda, you pretty much just see the sky (although everything works fine in the menus, just gets messed up in game

**Comment by gonetz** [comment]
Well, at least it shows something. Now we need to find, what causes these problems.
First, let's check that it is depth buffer problem. Disable depth compare: in OGLRender::_updateStates find 
`if (config.frameBufferEmulation.N64DepthCompare) `
You may comment the whole if. Insert these two lines after it:
```
		glDisable( GL_DEPTH_TEST );
		glDepthMask( FALSE );
```
You should see most of geometry now, incorrect of course.

If you get expected result, the problem is with depth buffer indeed. The depth buffer either not cleared properly or depth compare functions work incorrect.

**Comment by loganmc10** [comment]
This is what it looks like with those changes:

![snapshot](https://cloud.githubusercontent.com/assets/848146/15681130/68a30034-2715-11e6-9347-2a3bde9ad341.png)

**Comment by gonetz** [comment]
I'm not sure that it is correct,  I don't have quick save from this place. Let's check something simple: here how SM64 looks on my PC with depth compare disabled:
![super_mario_64-000](https://cloud.githubusercontent.com/assets/8983092/15704284/d11b63b2-280b-11e6-87c7-945a63479288.png)

**Comment by loganmc10** [comment]
The Zelda screenshot is not correct, you can see through Links shield, and the guard in the distance looks like he is floating

**Comment by gonetz** [comment]
Graphics glitches are expected because depth compare is disabled. I mean, I don't know how it should look because I don't have save from this place and and can't check it on my system. Mario is a good test because you don't need a save to get there.

**Comment by loganmc10** [comment]
So this is how it looks on the Raspberry Pi when I comment out that stuff in OGLRender::_updateStates and add glDisable( GL_DEPTH_TEST ); and glDepthMask( FALSE );

![snapshot](https://cloud.githubusercontent.com/assets/848146/15714343/68dd1dce-27d7-11e6-9edb-7505886f87ef.png)

**Comment by loganmc10** [comment]
Hey sorry I goofed, I had Framebuffer Emulation disabled when I took that screenshot... It looks like this with FB Emulation enabled.
![snapshot](https://cloud.githubusercontent.com/assets/848146/15718358/3d3620c4-27e8-11e6-8844-26334d8421a8.png)

However that Zelda screenshot was taken with Framebuffer Emulation enabled.

**Comment by gonetz** [comment]
Hmm, I saw this bug on Android. Do save state (F5) on this place and load from it (F7)

**Comment by gonetz** [comment]
I suppose that the glitches you have caused by wrong depth buffer. I can't guess, what exactly is wrong. Would be great to know, that your hardware capable to handle FBO properly. You may try to find some open-source demo to check it. I may only do some blind assumptions:
* suppose depth buffer is not clear properly. Plugin does not call glClearDepth function, because the initial value must be 1, which is fine. Probably on your system initial value did not set. You may call glClearDepth(1) in OGLRender::_initStates(). Also make sure that OGLRender::clearDepthBuffer is called with no errors (put isGLError() call at the end).
* probably some depth compare modes do not work. Remove your changes in OGLRender::_updateStates , then after
```
switch (gDP.otherMode.depthMode) {
...
}
```
add
```
glDisable(GL_POLYGON_OFFSET_FILL);
glDepthFunc(GL_LESS);
```
to force depth mode, which definitely must work.

**Comment by loganmc10** [comment]
I am starting to believe that the SDL2 video driver for the Raspberry Pi is incomplete. Based on my reading it seems that EGL_DEPTH_SIZE needs to be set for the depth buffer to work properly, but I don't think that SDL2 is setting that at all, so it defaults to 0. I could be wrong, I'll do some more research, but it seems like the SDL2 driver is incomplete.

**Comment by loganmc10** [comment]
None of those changes make any difference, the output is exactly the same. I'll do some more digging, I don't want to waste any of your time, I'm going to look into the SDL2 video driver a bit, perhaps something is wrong with the EGL context it is setting up. I'm not really sure...

**Comment by dankcushions** [comment]
@loganmc10 are you using retropie? i believe it uses a SDL2 fork - see https://github.com/RetroPie/SDL-mirror - which is a fork of another mirror. maybe something is updated further up...?

perhaps @gizmo98 might know something?

**Comment by loganmc10** [comment]
@dankcushions No I'm not using retropie, I compile SDL2 and mupen64plus/GlideN64 from source, I'm the 320x240 SDL2 stretching guy from Reddit...

If someone could try a different GLES2 device (like an Android phone) with the depth_renderbuffer and -DUSE_DEPTH_RENDERBUFFER enabled that would help, since it would determine if the issue is with the Raspberry Pi or the renderbuffer code.

@gizmo98 @fzurita

**Comment by fzurita** [comment]
Yeah, sure, can try that later today.

**Comment by gizmo98** [comment]
We use SDL 2.0.4 with minor modifications.
https://github.com/RetroPie/SDL-mirror/commits/retropie-2.0.4?author=joolswills

I have not tested if https://github.com/gonetz/GLideN64/commit/d5bc5d0d3051a6d2152856a51d0bd954312ed3c5 runs better with our sdl version.

**Comment by fzurita** [comment]
The code straight out of depth_renderbuffer branch gives me a black screen (when defining USE_DEPTH_RENDERBUFFER )

**Comment by loganmc10** [comment]
@fzurita Can you go into the FBOTextureFormats::init (OpenGL.cpp) and change
```
depthInternalFormat = GL_DEPTH_COMPONENT;
```

to
```
depthInternalFormat = GL_DEPTH_COMPONENT16
```

I don't think it likes it without the "16";

**Comment by fzurita** [comment]
Ok, that worked. Also, I don't get the depth issues you are seeing. Mario's face at the start of Mario 64 looks ok. Zelda's shield also looks ok.

**Comment by loganmc10** [comment]
So once you change depthInternalFormat everything looks normal? I hate the Pi...

**Comment by fzurita** [comment]
Yep, once I changed the ```depthInternalFormat``` and ```depthFormatBytes```, everything looked ok. Sounds like the Pi has some non complaint drivers.

**Comment by loganmc10** [comment]
Ya it's strange because I don't get any glErrors or anything kind of error whatsoever, it just looks all messed up

**Comment by fzurita** [comment]
Android also has buggy drivers or drivers that behave differently from GPU to GPU. Adreno has only started working well for the most part after I found several fixes that were specific to that GPU.

I have a PowerVR driver (the Nexus Player) and it has a ton of issues and it's generally sluggish overall. I'm trying to find the causes of the slow down for that device right now...

The only device that has worked well right off the bat for me has been the NVidia Shield TV. The NVidia Shield TV even has full OpenGL support which is nice.

**Comment by gonetz** [comment]
@fzurita So, depthInternalFormat = GL_DEPTH_COMPONENT16 is not Pi specific and must be enabled for GLES2 in general, right?

**Comment by fzurita** [comment]
It was only needed after I defined -DUSE_DEPTH_RENDERBUFFER. Before I defined that, everything rendered correctly with GL_DEPTH_COMPONENT.

I could attempt to use GL_DEPTH_COMPONENT16 and see what happens in the master branch. It would be nice to get rid of an offer.

**Comment by gonetz** [comment]
Ok, please do it.

**Comment by fzurita** [comment]
Ok, I see no issues in the master branch when using GL_DEPTH_COMPONENT16 under GLES 2.0.

Also, I'm able to reproduce the issue in Mario's face in the Mario 64 intro in the master branch with no changes. It only happens if you start the game fresh. If you load a save state to Mario's face, the issue doesn't happen.

**Comment by gonetz** [comment]
Thanks!

**Comment by gonetz** [comment]
@loganmc10 may be close it as won't fix?

**Comment by loganmc10** [comment]
@gonetz yes, I'm hoping it's a duplicate of https://github.com/gonetz/GLideN64/issues/1011, but I've hit a roadblock with this so it can be closed
